### ISSUE TITLE
PHP 8.4 packages

### ISSUE THREAD

**Issue by oerdnj** [oerdnj/deb.sury.org#2176]
This is a tracking issue for PHP 8.4 status.

* [x] PHP 8.4
  * [x] Create a new package for PHP 8.4.0alpha1
  * [x] Update the package with PHP 8.4.0alpha2 (when available)
  * [x] Upload to php-qa repositories
  * [x] Upload to Debian experimental
  * [x] Initiate transition for Debian trixie (RC version)
  * [ ] Upload PHP 8.4 version to Debian unstable (RC version)
  * [ ] Upload final PHP 8.4 version to PHP repositories
  * [ ] Upload final PHP 8.4 version to Debian unstable
* [ ] php-defaults
  * [x] Set PHP 8.4 as single supported version in debian/experimental branch
  * [x] Upload to php-qa repositories
  * [x] Remove IMAP and pspell extensions from the modulelist
  * [x] Upload to Debian experimental
  * [ ] Add PHP 8.4 for packages.sury.org version (along with PHP 5.6 - PHP 8.3)
  * [ ] Upload multiversion php-defaults to PHP repositories
* [x] Unbundled IMAP extension
  * [x] Create a new package for unbundled IMAP extension
  * [x] Upload to php-qa repositories
  * [x] Upload to Debian experimental (after PHP 8.4 passes NEW)
* [x] Unbundled pspell extension
  * [x] Create a new package for unbundled pspell extension
  * [x] Upload to php-qa repositories
  * [x] Upload to Debian experimental (after PHP 8.4 passes NEW)
* [ ] Current PECL extensions
  * [x] Test all current PECL extensions against PHP 8.4
  * [ ] Build status of PECL extensions
    * [x] libvirt-php
    * [x] php-amqp
    * [x] php-apcu
    * [x] php-ast
    * [ ] php-decimal (missing libmpdec-dev on noble)
    * [x] php-dio
    * [x] php-ds
    * [x] php-excimer
    * [x] php-gearman
    * [x] php-geoip
    * [x] php-gmagick
    * [x] php-gnupg
    * [x] php-grpc
    * [x] php-igbinary
    * [x] php-imagick
    * [x] php-imap
    * [x] php-inotify
    * [x] php-lz4 (will need manual debian/control.in update)
    * [x] php-mailparse
    * [x] php-maxminddb
    * [x] php-mcrypt
    * [x] php-memcache
    * [x] php-memcached
    * [x] php-mongodb
    * [x] php-msgpack
    * [x] php-oauth
    * [ ] php-openswoole
    * [x] php-opentelemetry
    * [x] php-pcov
    * [x] php-pecl-http
    * [ ] php-pinba (will need manual debian/control.in update)
    * [ ] php-phalcon5
    * [x] php-pq
    * [x] php-protobuf
    * [x] php-ps
    * [x] php-pspell
    * [x] php-psr
    * [x] php-radius
    * [x] php-raphf
    * [x] php-rdkafka
    * [x] php-redis (fatal error: standard/php_random.h: No such file or directory)
    * [x] php-rrd
    * [x] php-smbclient
    * [ ] php-solr (syntax error near unexpected token `-fprofile-arcs,' in configure)
    * [x] php-ssh2
    * [x] php-stomp
    * [ ] php-swoole
    * [x] php-uopz
    * [x] php-uploadprogress
    * [x] php-uuid
    * [x] php-vips
    * [x] php-xhprof
    * [x] php-xmlrpc
    * [x] php-yac
    * [x] php-yaml
    * [x] php-zmq
    * [x] php-zstd
    * [x] xdebug (3.4.0~alpha1 works)
  * [x] Not supported
    * [x] php-apcu-bc
    * [x] php-geoip
    * [x] php-lua
    * [x] php-mongo
    * [x] php-mysqlnd-ms
    * [x] php-propro
    * [x] php-radius
    * [x] php-sodium
    * [x] xcache
  * [ ] Remove
    * [ ] php-facedetect (remove completely?)
    * [ ] tideways (obsoleted by upstream, remove completely)
  * [x] Upload to Debian experimental (after PHP 8.4 passes NEW)
    * [x] libvirt-php
    * [x] php-amqp
    * [x] php-apcu
    * [x] php-ast
    * [ ] php-decimal (missing libmpdec-dev on noble)
    * [x] php-ds
    * [x] php-excimer (fatal error: ext/standard/php_mt_rand.h: No such file or directory)
    * [x] php-gearman
    * [x] php-gmagick
    * [x] php-gnupg
    * [ ] php-grpc
    * [ ] php-igbinary
    * [x] php-imagick
    * [x] php-imap
    * [ ] php-inotify
    * [x] php-mailparse
    * [x] php-maxminddb
    * [x] php-mcrypt (fatal error: ext/standard/php_rand.h: No such file or directory)
    * [x] php-memcache
    * [ ] php-memcached
    * [x] php-mongodb
    * [x] php-msgpack
    * [ ] php-oauth (fatal error: ext/standard/php_rand.h: No such file or directory)
    * [ ] php-openswoole
    * [ ] php-pcov (error: too many arguments to function ‘php_pcre_match_impl’)
    * [x] php-pecl-http (fatal error: ext/standard/php_lcg.h)
    * [ ] php-pinba (up to PHP 8.2 only)
    * [ ] php-pq
    * [x] php-pspell
    * [x] php-psr (tests failing)
    * [x] php-raphf
    * [ ] php-rdkafka
    * [ ] php-redis (fatal error: standard/php_random.h: No such file or directory)
    * [x] php-rrd
    * [ ] php-smbclient
    * [ ] php-solr (syntax error near unexpected token `-fprofile-arcs,' in configure)
    * [x] php-stomp
    * [ ] php-swoole
    * [ ] php-uopz
    * [x] php-uploadprogress
    * [x] php-uuid
    * [ ] php-vips
    * [ ] php-xhprof
    * [x] php-xmlrpc
    * [x] php-yac
    * [ ] php-yaml
    * [ ] php-zmq
    * [ ] php-zstd
    * [x] xdebug (usually need a new upstream version)

NOTE: As of now, I am not yet interested in bug reports until this tracking issue is finished.

**Comment by rfay** [comment]
Thanks so much for all the work on this! 

You say 
> Upload to Debian experimental

I'm not familiar with that, is it publicly available? Thanks!

**Comment by oerdnj** [comment]
> > Upload to Debian experimental
> 
> I'm not familiar with that, is it publicly available? Thanks!

https://wiki.debian.org/DebianExperimental

It's used to prepare the transition for Debian unstable (and other experiments).

**Comment by AnnoyingTechnology** [comment]
Any idea when APCu would be made available ?
(And thanks for all the work!)

**Comment by rfay** [comment]
Although they're listed above, I don't see php8.4-apcu php8.4-opcache php8.4-redis php8.4-uploadprogress php8.4-xdebug php8.4-xhprof (Debian 12, both arm64 and amd64)

Maybe I misunderstand the list. I thought "Current pecl extensions" that were checked were available now, but I must not understand it.

**Comment by Sn0wCrack** [comment]
Additionally for imagick, mailparse and phalcon for PHP 8.4 the following warnings is generated when installed:

```
PHP Warning:  PHP Startup: Unable to load dynamic library 'imagick.so' (tried: /usr/lib/php/20240924/imagick.so (/usr/lib/php/20240924/imagick.so: undefined symbol: php_strtolower), /usr/lib/php/20240924/imagick.so.so (/usr/lib/php/20240924/imagick.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0

PHP Warning:  PHP Startup: Unable to load dynamic library 'mailparse.so' (tried: /usr/lib/php/20240924/mailparse.so (/usr/lib/php/20240924/mailparse.so: undefined symbol: php_strtolower), /usr/lib/php/20240924/mailparse.so.so (/usr/lib/php/20240924/mailparse.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0

PHP Warning:  PHP Startup: Unable to load dynamic library 'phalcon.so' (tried: /usr/lib/php/20240924/phalcon.so (/usr/lib/php/20240924/phalcon.so: undefined symbol: php_strtoupper), /usr/lib/php/20240924/phalcon.so.so (/usr/lib/php/20240924/phalcon.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0
```

**Comment by Krinkle** [comment]
php8.4-apcu requires an upstream release of commits at https://github.com/krakjoe/apcu first since php-apcu 5.1.23 from PECL doesn't compile cleanly against PHP 8.4.

**Comment by DoobleD** [comment]
Thank you fo the work here!

Would it be possible to have php8.4-pgsql come with a pdo_pgsql compiled with libpq 17? That would allow the postgresql PDO to work with connection poolers such as pgbouncer in transaction mode. See [this](https://github.com/php/php-src/issues/12940#issuecomment-2451680438) issue.

I haven't seen php8.4-pgsql mentionned in this issue though so perhaps I'm not asking in the right place?

**Comment by oerdnj** [comment]
> Thank you fo the work here!
> 
> 
> 
> Would it be possible to have php8.4-pgsql come with a pdo_pgsql compiled with libpq 17? That would allow the postgresql PDO to work with connection poolers such as pgbouncer in transaction mode. See [this](https://github.com/php/php-src/issues/12940#issuecomment-2451680438) issue.

It gets compiled with whatever libraries are provided by the base system.  If you want libpq >= 17, you need to upgrade to Ubuntu/Debian that has libpq-dev >= 17

> 
> 
> I haven't seen php8.4-pgsql mentionned in this issue though so perhaps I'm not asking in the right place?

The list is only for PECL extensions, but generally speaking, you are asking in wrong place, you question is not directly relevant to the topic.

**Comment by rfay** [comment]
(I see new builds showing up for amd64 but not arm64. Are the arm64 builds failing?)

**Comment by DoobleD** [comment]
Thank you @oerdnj!

**Comment by BramDriesen** [comment]
Following 👀

**Comment by Salamandar** [comment]
No pressure here. I'm wondering when you plan to have a Debian 13 / Trixie repository up on packages.sury.org ? We at [YunoHost](https://yunohost.org) are starting to test our support for Trixie, to be ready when Trixie is released, while still using the Bookworm repo, so of course we have package conflicts :D

**Comment by oerdnj** [comment]
Not until we are in library transition freeze. It's too volatile until then.

**Comment by rfay** [comment]
Remaining ones that DDEV is missing (Bookworm, arm64)
`apcu redis uploadprogress xdebug xhprof xmlrpc`

I *think* I see those checked above, but maybe my eyes are skipping over something.

**Comment by Salamandar** [comment]
> Not until we are in library transition freeze. It's too volatile until then.

OK, i suppose I need to wait for https://release.debian.org/trixie/freeze_policy.html to display some dates then :D

**Comment by oerdnj** [comment]
Let me reiterate - until this issue is closed, I will just close bug reports against PHP 8.4 without any notice. This bug is prominent at the top and filling duplicate issue just means that people filling duplicate issues just haven't checked it.

**Comment by ariaieboy** [comment]
@oerdnj swoole 6.0.0-rc1 is compatible with PHP 8.4

**Comment by fretb** [comment]
Hi @oerdnj,

As always, thank you so much for your hard work you put into this.

Just an informative question, do you have any idea when PHP 8.4 extensions like imagick, memcached, ... will be available in your Debian repository? As they are available already in the Ubuntu PPA (e.g. http://ppa.launchpadcontent.net/ondrej/php/ubuntu/pool/main/p/php-imagick/php8.4-imagick_3.7.0-8+ubuntu20.04.1+deb.sury.org+1_amd64.deb)

For Debian e.g. https://packages.sury.org/php/pool/main/p/php-imagick/

Thanks!

**Comment by oerdnj** [comment]
I just fixed the arm64 builder, and I am waiting for all the PHP packages to finish building first. I’ll be expect that I’ll try to go through the PECL packages over weekend, try to rebuild and update the status on them.

**Comment by fretb** [comment]
Thanks for taking the time to respond!

**Comment by jokerisgod** [comment]
Hi @oerdnj 

Thanks for the hard work as always.

Please help with the below if possible.

PHP Warning:  PHP Startup: Unable to load dynamic library 'imagick.so' (tried: /usr/lib/php/20240924/imagick.so (/usr/lib/php/20240924/imagick.so: undefined symbol: php_strtolower), /usr/lib/php/20240924/imagick.so.so (/usr/lib/php/20240924/imagick.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0

**Comment by thbley** [comment]
Having pcov for debian bookworm would be great!

```
E: Unable to locate package php8.4-pcov
E: Couldn't find any package by glob 'php8.4-pcov'
E: Couldn't find any package by regex 'php8.4-pcov'

```

**Comment by oerdnj** [comment]
> NOTE: As of now, I am not yet interested in bug reports until this tracking issue is finished.

**Comment by thbley** [comment]
Here is my workaround for getting pcov running with debian and php 8.4:
(based on https://github.com/krakjoe/pcov/pull/111)

```
DEBIAN_FRONTEND=noninteractive apt-get -y --no-install-recommends install unzip php8.4-dev make
curl -sSL "https://github.com/krakjoe/pcov/archive/refs/heads/issue-php84.zip" -o pcov.zip
unzip -q pcov.zip
rm pcov.zip
cd pcov-issue-php84
phpize
./configure --enable-pcov
make
make install
echo "extension=pcov.so" > /etc/php/8.4/cli/conf.d/20-pcov.ini
cd ..
rm -rf pcov-issue-php84
apt-get purge -y --auto-remove php8.4-dev make
apt-get clean
```

**Comment by oerdnj** [comment]
> I just fixed the arm64 builder, and I am waiting for all the PHP packages to finish building first. I’ll be expect that I’ll try to go through the PECL packages over weekend, try to rebuild and update the status on them.

There was some troubles getting the extensions rebuilt, but they should resolve themselves as 7.3, 7.4 and 8.0 got rebuilt.

**Comment by oerdnj** [comment]
Hey @derickr, do you have any ETA for final xdebug 3.4.0? Should I prepare interim xdebug 3.4.0~beta1 packages or is it matter of small number of days?

**Comment by derickr** [comment]
Was hoping for today but then I got stuck in an airport. It'll be in the next few days.

**Comment by oerdnj** [comment]
> Was hoping for today but then I got stuck in an airport.

Aw, sorry to hear that. Hope you at least get some compensation.

> It'll be in the next few days.

Thanks, then it can wait a bit.

**Comment by onlime** [comment]
Great @derickr & @oerdnj thanks for releasing all extensions so fast, now even Xdebug 3.4.0 👏🙏

```bash
$ apt show php8.4-xdebug
Package: php8.4-xdebug
Version: 3.4.0-1+0~20241202.61+debian12~1.gbped756b
```

**Comment by stasadev** [comment]
@oerdnj, thank you for packaging Xdebug 3.4.0.

I'm not sure if this is related to packaging, but Xdebug shows a JIT warning:

```
$ apt show php8.4-xdebug
Package: php8.4-xdebug
Version: 3.4.0-1+0~20241202.61+debian12~1.gbped756b

$ php -v
PHP Warning:  JIT is incompatible with third party extensions that override zend_execute_ex(). JIT disabled. in Unknown on line 0

Warning: JIT is incompatible with third party extensions that override zend_execute_ex(). JIT disabled. in Unknown on line 0
PHP 8.4.1 (cli) (built: Nov 25 2024 18:23:27) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.4.1, Copyright (c) Zend Technologies
    with Zend OPcache v8.4.1, Copyright (c), by Zend Technologies
    with Xdebug v3.4.0, Copyright (c) 2002-2024, by Derick Rethans
```

We are currently "fixing" this with `opcache.jit=disable` as recommended in https://bugs.xdebug.org/view.php?id=2293

**Comment by derickr** [comment]
It is opcache that shows this warning. Not Xdebug.

**Comment by rfay** [comment]
Great stuff @oerdnj , thanks so much. All of DDEV's standard packages are now working on PHP 8.4 🚀

**Comment by rikgirbes** [comment]
I'm encountering some issues with certain extensions while trying to install PHP 8.4. Specifically, I can't get `gd`, `redis`, and `mbstring` to install. I noticed there's a problem listed for `redis`, but I don't see `mbstring` or `gd` mentioned at all. Am I missing something?

**Comment by oerdnj** [comment]
The packages have been uploaded to Debian unstable now and while there might be some loose ends here and there, I think the PHP 8.4 support is now official.

**Comment by dada-amater** [comment]
Hi @oerdnj , at first, thank you for your great work.

We use ubuntu 24.04 and we are not able to install `php8.4-amqp`:
> Package 'php8.4-amqp' has no installation candidate

All other extensions are available. Is there anything we can do about it? Thank you.

**Comment by oerdnj** [comment]
> We use ubuntu 24.04 and we are not able to install php8.4-amqp:

I would prefer a complete information, so I don't have to play guessing game. There was a failed arm64 build, but I have absolutely no idea where restarting the build will help you or you have other problem because I can't see your system.

> Is there anything we can do about it?

You should fill a new issue and fill out all the details that I ask in the issue template.

**Comment by oerdnj** [comment]
I am now going to lock this issue, so I don't get these random comments since I need complete information and not just "foo doesn't work", so I don't have to spend time figuring out what foo actually means.
